lib/internal/buffer.js

```diff
@@ -1,13 +1,797 @@
 'use strict';
 
 const binding = process.binding('buffer');
+const { TypeError, RangeError } = require('internal/errors');
 const { setupBufferJS } = binding;
 
 // Remove from the binding so that function is only available as exported here.
 // (That is, for internal use only.)
 delete binding.setupBufferJS;
 
+// Temporary buffers to convert numbers.
+const float32Array = new Float32Array(1);
+const uInt8Float32Array = new Uint8Array(float32Array.buffer);
+const float64Array = new Float64Array(1);
+const uInt8Float64Array = new Uint8Array(float64Array.buffer);
+
+// Check endianness.
+float32Array[0] = -1;
+const bigEndian = uInt8Float32Array[3] === 0;
+
+function checkBounds(buf, offset, byteLength) {
+  checkNumberType(offset);
+  if (buf[offset] === undefined || buf[offset + byteLength] === undefined)
+    boundsError(offset, buf.length - (byteLength + 1));
+}
+
+function checkInt(value, min, max, buf, offset, byteLength) {
+  if (value > max || value < min) {
+    throw new RangeError('ERR_OUT_OF_RANGE',
+                         'value', `>= ${min} and <= ${max}`, value);
+  }
+  checkBounds(buf, offset, byteLength);
+}
+
+function checkNumberType(value, type) {
+  if (typeof value !== 'number') {
+    throw new TypeError('ERR_INVALID_ARG_TYPE',
+                        type || 'offset', 'number', value);
+  }
+}
+
+function boundsError(value, length, type) {
+  if (Math.floor(value) !== value) {
+    checkNumberType(value, type);
```

Comment:
That's a little confused. In other constructions where `type` is used `offset` is default value.
Meanwhile `checkNumberType` (currently `validateNumber`) will throw `AssertionError` if `type` is undefined, because `ERR_INVALID_ARG_TYPE` have assert which check that `type` should be a `string`.
Should it be changed to: `function boundsError(value, length, type = 'offset') {` ?
@BridgeAR

Reply:
The default value is set in `checkNumberType` in this PR. Seems like it regressed in #24815. I'll open a PR to fix this. Thanks for reporting it!

Reply:
I just checked this again and all call sites seem fine, even with the PR I mentioned. Do you have any case that actually throws an `AssertionError`? I can't come up with any case.

Reply:
There is no way to receive `AssertionError` right now. It's only about readability, when I checked code and found that `type` passed as `undefined` I asked myself "is it possible that some error will be thrown?", then I found `assert` and start looked for real case when error can be thrown. But then I realized that `validateNumber` will never throw error here because always called before.
If default value will be exists probably this will be easier for reading.

Reply:
What about such change? Or will be reject because not fix anything?
```js
diff --git a/lib/internal/buffer.js b/lib/internal/buffer.js
index f3bc0e48d1..5a6dddf9f8 100644
--- a/lib/internal/buffer.js
+++ b/lib/internal/buffer.js
@@ -60,17 +60,18 @@ function checkInt(value, min, max, buf, offset, byteLength) {
   checkBounds(buf, offset, byteLength);
 }
 
-function boundsError(value, length, type) {
+function boundsError(value, length, type = 'offset') {
   if (Math.floor(value) !== value) {
     validateNumber(value, type);
-    throw new ERR_OUT_OF_RANGE(type || 'offset', 'an integer', value);
+    throw new ERR_OUT_OF_RANGE(type, 'an integer', value);
   }
 
   if (length < 0)
     throw new ERR_BUFFER_OUT_OF_BOUNDS();
 
-  throw new ERR_OUT_OF_RANGE(type || 'offset',
-                             `>= ${type ? 1 : 0} and <= ${length}`,
+  const lowerBound = type === 'offset' ? 0 : 1;
+  throw new ERR_OUT_OF_RANGE(type,
+                             `>= ${lowerBound} and <= ${length}`,
                              value);
 }
 ```